File: ts/tests/gen-data.test.ts
/** The generator is part of CI, so its drift rates are asserted like any other behaviour. */
import assert from "node:assert/strict";
import { mkdtempSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { test } from "node:test";

import { compare } from "../src/engine.ts";
import { makeOptions } from "../src/types.ts";
import { COLUMNS, genData, parseRows } from "../scripts/gen-data.ts";

const KEY = ["account_id", "txn_id"];

function gen(rows = 10_000) {
  const dir = mkdtempSync(join(tmpdir(), "csvdiff-gen-"));
  const a = join(dir, "a.csv");
  const b = join(dir, "b.csv");
  genData(rows, a, b);
  return { a, b };
}

test("parseRows", () => {
  assert.equal(parseRows("10k"), 10_000);
  assert.equal(parseRows("1m"), 1_000_000);
  assert.equal(parseRows("2.5M"), 2_500_000);
  assert.equal(parseRows("1,000"), 1000);
  assert.equal(parseRows("42"), 42);
  assert.equal(COLUMNS.length, 20);
});

test("shape and drift", async () => {
  const { a, b } = gen();
  const r = await compare(a, b, makeOptions({ key: KEY, ignore: ["updated_at"] }));
  const c = r.counts;
  assert.equal(r.meta.compared.length, 17); // 20 columns - 2 key - 1 ignored
  assert.equal(c.a_keys, 10_000);
  assert.equal(c.b_keys, 10_000);
  assert.equal(c.added, 10); // 0.1% each
  assert.equal(c.removed, 10);
  assert.equal(c.a_dup_keys, 1);
  assert.equal(c.b_dup_keys, 1);
  const ratio = c.changed / c.matched;
  assert.ok(ratio > 0.055 && ratio < 0.065, `~6% of matched rows differ, got ${ratio}`);
  const by = Object.fromEntries(r.columns.map((x) => [x.name, x]));
  assert.ok(by["status"]!.changed > by["amount"]!.changed && by["amount"]!.changed > 0);
  assert.equal(by["value_date"]!.blanked, by["value_date"]!.changed);
  assert.ok(by["value_date"]!.changed > 0);
  assert.equal(by["note"]!.changed, 0); // untouched columns stay untouched
});

test("deterministic", () => {
  const one = gen(2_000);
  const two = gen(2_000);
  assert.ok(readFileSync(one.a).equals(readFileSync(two.a)));
  assert.ok(readFileSync(one.b).equals(readFileSync(two.b)));
});

test("money never passes through a float", () => {
  // Two-decimal columns come from integer cents, so every implementation of the
  // generator writes the same digits whatever its rounding rule.
  const { a, b } = gen(2_000);
  for (const path of [a, b]) {
    const lines = readFileSync(path, "utf-8").split("\n").slice(1, 500).filter(Boolean);
    const at = (line: string, name: string) => line.split(",")[COLUMNS.indexOf(name)]!;
    for (const line of lines) {
      for (const column of ["amount", "fee", "balance"]) {
        assert.match(at(line, column), /^-?\d+\.\d{2}$/, `${column} is not 2dp in ${line}`);
      }
      assert.match(at(line, "rate"), /^0\.\d{4}$/);
    }
  }
});

File: java/src/test/java/dev/csvdiff/GenDataTest.java
package dev.csvdiff;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import dev.csvdiff.bench.GenData;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * The generator is part of CI, so its drift recipe is asserted like any other behaviour. Changing a
 * rate here means updating this test deliberately, not to make it pass.
 */
class GenDataTest {

  private static final List<String> KEY = List.of("account_id", "txn_id");

  @Test
  @DisplayName("row counts and drift rates match the documented recipe")
  void shapeAndDrift(@TempDir Path dir) throws Exception {
    Path a = dir.resolve("a.csv");
    Path b = dir.resolve("b.csv");
    GenData.generate(10_000, a, b, 7);

    var result =
        CsvDiff.compare(a, b, Options.builder().key(KEY).ignore(List.of("updated_at")).build());
    var c = result.counts();

    assertEquals(17, result.meta().compared().size(), "20 columns - 2 key - 1 ignored");
    assertEquals(10_000, c.aKeys());
    assertEquals(10_000, c.bKeys());
    assertEquals(10, c.added(), "0.1% of rows are only in B");
    assertEquals(10, c.removed(), "0.1% of rows are only in A");
    assertEquals(1, c.aDupKeys());
    assertEquals(1, c.bDupKeys());

    double ratio = (double) c.changed() / c.matched();
    assertTrue(ratio > 0.055 && ratio < 0.065, "about 6% of matched rows differ, got " + ratio);

    Map<String, Contract.ColumnStat> by =
        result.columns().stream().collect(Collectors.toMap(Contract.ColumnStat::name, s -> s));
    assertTrue(by.get("status").changed() > by.get("amount").changed());
    assertTrue(by.get("amount").changed() > 0);
    assertEquals(by.get("value_date").changed(), by.get("value_date").blanked());
    assertTrue(by.get("value_date").changed() > 0);
    assertEquals(0, by.get("note").changed(), "untouched columns stay untouched");
  }

  @Test
  @DisplayName("the same seed produces byte-identical files")
  void deterministic(@TempDir Path dir) throws Exception {
    Path a1 = dir.resolve("a1.csv");
    Path b1 = dir.resolve("b1.csv");
    Path a2 = dir.resolve("a2.csv");
    Path b2 = dir.resolve("b2.csv");
    GenData.generate(2_000, a1, b1, 7);
    GenData.generate(2_000, a2, b2, 7);

    assertArrayEquals(Files.readAllBytes(a1), Files.readAllBytes(a2));
    assertArrayEquals(Files.readAllBytes(b1), Files.readAllBytes(b2));
  }

  @Test
  @DisplayName("row-count shorthands parse")
  void parseRows() {
    assertEquals(10_000, GenData.parseRows("10k"));
    assertEquals(1_000_000, GenData.parseRows("1m"));
    assertEquals(2_500_000, GenData.parseRows("2.5M"));
    assertEquals(1_000, GenData.parseRows("1,000"));
    assertEquals(42, GenData.parseRows("42"));
    assertEquals(20, GenData.COLUMNS.size());
  }
}
